docs/RFCS/20220628_invisible_index.md

```diff
@@ -0,0 +1,212 @@
+
+```
+   ALTER TABLE ... ADD [COLUMN] [IF NOT EXISTS] <colname> <type> [<constraint...>]
+   ALTER TABLE ... ADD <constraint>
+   ALTER TABLE ... ALTER PRIMARY KEY USING INDEX <name>
+```
+
+<blockquote><details>
+<summary>constraint</summary>
+
+```
+[CONSTRAINT <constraintname>] {NULL | NOT NULL | UNIQUE [WITHOUT INDEX | *** WITH INVISIBLE INDEX  | WITH NOT VISIBLE INDEX *** ]| PRIMARY KEY [*** WITH INVISIBLE INDEX  | WITH NOT VISIBLE INDEX *** ]| CHECK (<expr>) | DEFAULT <expr>}
+```
+</blockquote></details>
+
+[//]: # (Alter Index)
+4. `Alter Index`
+```
+ALTER INDEX [IF EXISTS] <idxname> [INVISIBLE | NOT VISIBLE]
+```
+
+[//]: # (Show Constraint)
+5. `Show Constraint`
+```
+SHOW CONSTRAINT FROM table_name with_comment
+```
+```
+table_name  constraint_name  constraint_type  details(*** add invisible index details ***)  validated
+```
+
+[//]: # (Show Index)
+6. A new column needs to be added to `crdb_internal.table_indexes`.
+```
+descriptor_id  descriptor_name  index_id  index_name  index_type  is_unique  is_inverted  is_sharded *** is_hidden | is_invisible | is_not_visible *** shard_bucket_count created_at
+```
+
+7. A new column needs to be added to the output of following SQL statements:
+``` 
+SHOW INDEX FROM (table_name) 
+SHOW INDEXES FROM(table_name) 
+SHOW KEYS FROM (table_name) 
+```
+```
+table_name  index_name  non_unique  seq_in_index  column_name  direction  storing  implicit *** is_hidden | is_invisible | is_not_visible ***  
+```
+
+```
+SHOW INDEX FROM DATABASE(database_name) 
+SHOW INDEXES FROM DATABASE (database_name) 
+SHOW KEYS FROM DATABASE (database_name) 
+```
+```
+table_name  index_name  non_unique  seq_in_index  column_name  direction  storing  implicit *** is_hidden | is_invisible | is_not_visible ***  
+```
+
+7. Note that `CREATE CONSTRAINT` and `ALTER CONSTRAINT` are both not supported by the parser.
+
+### Discussion 
+CockroachDB currently supports invisible column feature. For this
+feature, `NOT VISIBLE` is used for its SQL statement, and `is_hidden` is used
+for the new column added to `SHOW INDEX`. It would be nice to stay consistent.
+But [MySQL](https://dev.mysql.com/doc/refman/8.0/en/invisible-indexes.html) and
+[Oracle](https://oracle-base.com/articles/11g/invisible-indexes-11gr1) both
+support `INVISIBLE` instead.
+[MySQL](https://dev.mysql.com/doc/refman/8.0/en/invisible-columns.html) and
+[Oracle](https://oracle-base.com/articles/12c/invisible-columns-12cr1) both use
+`INVISIBLE` for their invisible index and invisible column feature.
+[MySQL](https://dev.mysql.com/doc/refman/8.0/en/invisible-indexes.html) use
+`is_visible` for the new column added to `SHOW INDEX`.
+[Oracle](http://www.dba-oracle.com/t_11g_new_index_features.htm) use
+`VISIBILITY` for the new column added to `SHOW INDEX`.
```

Comment:
i kind of wonder whether INVISIBLE / NOT VISIBLE is the right way to "express" this. My understanding is you still have to pay the write penalty of the indexes, it just doesn't get "used" (which also makes me wonder whether this feature is useful in the ADD case - see my motivation comment above). `NOT USED` sounds a bit better semantically to me. but other databases have made it INVISIBLE so maybe we're ok.

this is not a blocking comment and i can easily be convinced otherwise.